lib/phase/merge.js

```diff
@@ -0,0 +1,113 @@
+    const templateTag = await pipelineTemplateTagFactory.get({
+        name,
+        namespace,
+        tag: versionOrTag
+    });
+    const version = templateTag ? templateTag.version : versionOrTag;
+    const pipelineTemplate = await pipelineTemplateVersionFactory.getWithMetadata(
+        {
+            name,
+            namespace,
+            version
+        },
+        pipelineTemplateFactory
+    );
+
+    return pipelineTemplate;
+}
+
+/**
+ * Merge parsed yaml setting with pipeline template setting
+ * Pipeline yaml will take precedence over pipeline template
+ * @method mergeSetting
+ * @method
+ * @param   {Object}        parsedDoc              Parsed screwdriver yaml
+ * @param   {Object}        pipelineTemplate       Pipeline template
+ * @return  {Object}                               Merged yaml settings
+ */
+function mergeSetting(parsedDoc, pipelineTemplate) {
+    const newPipeline = pipelineTemplate.config;
+    const yamlSharedConfig = parsedDoc.shared || {};
+    const templateSharedConfig = newPipeline.shared || {};
+
+    templateSharedConfig.environment = Hoek.merge(templateSharedConfig.environment || {}, {
+        SD_PIPELINE_TEMPLATE_FULLNAME: parsedDoc.template,
+        SD_PIPELINE_TEMPLATE_NAME: pipelineTemplate.name,
+        SD_PIPELINE_TEMPLATE_NAMESPACE: pipelineTemplate.namespace,
+        SD_PIPELINE_TEMPLATE_VERSION: pipelineTemplate.version
+    });
+
+    templateSharedConfig.environment = Hoek.merge(templateSharedConfig.environment, yamlSharedConfig.environment || {});
+    templateSharedConfig.settings = Hoek.merge(templateSharedConfig.settings || {}, yamlSharedConfig.settings || {});
+
+    if (yamlSharedConfig.image || templateSharedConfig.image) {
+        templateSharedConfig.image = yamlSharedConfig.image || templateSharedConfig.image;
+    }
+
+    newPipeline.shared = templateSharedConfig;
+
+    return newPipeline;
+}
+
+/**
+ * Merge Phase
+ *
+ * Merge pipeline template with screwdriver yaml
+ * @method
+ * @param   {Object}                          parsedDoc                                   Document that went through structural parsing
+ * @param   {PipelineTemplateVersionFactory}  config.pipelineTemplateVersionFactory       PiplineTemplateVersion Factory to get pipeline templates
+ * @param   {PipelineTemplateTagFactory}      config.pipelineTemplateTagFactory           PipelineTemplateTag Factory to get pipeline templates tag
+ * @param   {PipelineTemplateFactory}         config.pipelineTemplateFactory              PipelineTemplate Factory to get pipeline templates
+ * @returns {Promise}
+ */
+module.exports = async (
+    parsedDoc,
+    pipelineTemplateVersionFactory,
+    pipelineTemplateTagFactory,
+    pipelineTemplateFactory
+) => {
+    const warnings = [];
+    const pipelineTemplateName = Hoek.reach(parsedDoc, 'template');
+
+    if (!pipelineTemplateName) {
+        return Promise.resolve({ mergedDoc: parsedDoc, warnings });
```

Comment:
you don't need to use Promise.resolve({}) in an async function. The async keyword automatically wraps the return value of the function in a Promise. If you return an object, it will be equivalent to returning Promise.resolve({}).